log/levels/levels.go

```diff
@@ -0,0 +1,113 @@
+package levels
+
+import "github.com/go-kit/kit/log"
+
+// Levels provides a leveled logging wrapper around a logger. It has five
+// levels: debug, info, warning (warn), error, and critical (crit). If you
+// want a different set of levels, you can create your own levels type very
+// easily, and you can elide the configuration.
+type Levels struct {
+	ctx log.Context
+	cfg *config
+}
+
+// New creates a new leveled logger, wrapping the passed logger.
+func New(logger log.Logger, options ...Option) Levels {
+	cfg := &config{
+		levelKey:   "level",
+		debugValue: "debug",
+		infoValue:  "info",
+		warnValue:  "warn",
+		errorValue: "error",
+		critValue:  "crit",
+	}
+	for _, option := range options {
+		option(cfg)
+	}
+	return Levels{
+		ctx: log.NewContext(logger),
+		cfg: cfg,
+	}
+}
+
+// With returns a new leveled logger that includes keyvals in all log events.
+func (l Levels) With(keyvals ...interface{}) Levels {
+	return Levels{
+		ctx: l.ctx.With(keyvals...),
+		cfg: l.cfg,
+	}
+}
+
+// Debug logs a debug event along with keyvals.
+func (l Levels) Debug(keyvals ...interface{}) error {
+	return l.ctx.WithPrefix(l.cfg.levelKey, l.cfg.debugValue).Log(keyvals...)
+}
+
+// Info logs an info event along with keyvals.
+func (l Levels) Info(keyvals ...interface{}) error {
+	return l.ctx.WithPrefix(l.cfg.levelKey, l.cfg.infoValue).Log(keyvals...)
+}
+
+// Warn logs a warn event along with keyvals.
+func (l Levels) Warn(keyvals ...interface{}) error {
+	return l.ctx.WithPrefix(l.cfg.levelKey, l.cfg.warnValue).Log(keyvals...)
+}
+
+// Error logs an error event along with keyvals.
+func (l Levels) Error(keyvals ...interface{}) error {
+	return l.ctx.WithPrefix(l.cfg.levelKey, l.cfg.errorValue).Log(keyvals...)
+}
+
+// Crit logs a crit event along with keyvals.
+func (l Levels) Crit(keyvals ...interface{}) error {
+	return l.ctx.WithPrefix(l.cfg.levelKey, l.cfg.critValue).Log(keyvals...)
+}
```

Comment:
This seems to be heading back toward what we have in the master branch with a separate logger/context for each level. The downside to this approach is that each call to `Levels.With` must update all the leveled contexts. It seems difficult to justify the overhead of calling `With` on this many contexts because to break even the application must create more than five subsequent log events in the scope of the context.